pkg/capture/capture_manager.go

```diff
@@ -1,436 +1,355 @@
+			statusmapMutex.Lock()
+			statusmap[mc.iface] = *status
+			statusmapMutex.Unlock()
+		})
 	}
 	rg.Wait()
 
-	elapsed := time.Since(t0).Round(time.Millisecond)
-
-	logger.With("elapsed", elapsed.String()).Debug("updated interface list")
+	logger.With(
+		"elapsed", time.Since(t0).Round(time.Millisecond).String(),
+		"ifaces", ifaces,
+	).Debug("retrieved interface status")
 
+	return
 }
 
-// StateAll returns the states of all managed Capture instances (without touching
-// the capture stats)
-// func (cm *Manager) StateAll() map[string]capturetypes.State {
-// 	statemapMutex := sync.Mutex{}
-// 	statemap := make(map[string]capturetypes.State)
-
-// 	var rg RunGroup
-// 	for iface, mc := range cm.capturesCopy() {
-// 		iface, mc := iface, mc
-// 		rg.Run(func() {
-// 			mc.capture.mutex.Lock()
-// 			state := mc.capture.state
-// 			mc.capture.mutex.Unlock()
-// 			statemapMutex.Lock()
-// 			statemap[iface] = state
-// 			statemapMutex.Unlock()
-// 		})
-// 	}
-// 	rg.Wait()
-
-// 	return statemap
-// }
-
-// Status returns the statuses of all interfaces provided in the arguments
-func (cm *Manager) Status(ifaces ...string) map[string]capturetypes.InterfaceStatus {
-	statusmapMutex := sync.Mutex{}
-	statusmap := make(map[string]capturetypes.InterfaceStatus)
+// Update the configuration for all (or a set of) interfaces
+func (cm *Manager) Update(ctx context.Context, ifaces config.Ifaces) {
 
-	var rg RunGroup
-	cmCopy := cm.capturesCopy()
-
-	if len(ifaces) > 0 {
-		for _, iface := range ifaces {
-			iface := iface
-			mc, exists := cmCopy[iface]
-			if exists {
-				rg.Run(func() {
-					status := mc.capture.Status()
-					statusmapMutex.Lock()
-					statusmap[iface] = status
-					statusmapMutex.Unlock()
-				})
-			}
+	// Build set of interfaces to enable / disable
+	var (
+		ifaceSet                    = make(map[string]struct{})
+		enableIfaces, disableIfaces []string
+	)
+	for iface := range ifaces {
+		ifaceSet[iface] = struct{}{}
+		if _, exists := cm.captures.Get(iface); !exists {
+			enableIfaces = append(enableIfaces, iface)
 		}
-	} else {
-		for iface, mc := range cmCopy {
-			iface, mc := iface, mc
-			rg.Run(func() {
-				status := mc.capture.Status()
-				statusmapMutex.Lock()
-				statusmap[iface] = status
-				statusmapMutex.Unlock()
-			})
+	}
+	for iface := range cm.captures.Map {
+		if _, exists := ifaceSet[iface]; !exists {
+			disableIfaces = append(disableIfaces, iface)
 		}
 	}
-	rg.Wait()
 
-	return statusmap
+	cm.update(ctx, ifaces, enableIfaces, disableIfaces)
 }
 
-// ActiveFlows returns a copy of the current in-memory flow map. If iface is "all", flows for every interface are returned
-func (cm *Manager) ActiveFlows(ifaces ...string) map[string]capturetypes.FlowInfos {
-	ifaceFlows := make(map[string]*capturetypes.FlowLog)
-	ifaceFlowsMutex := sync.Mutex{}
+func (cm *Manager) update(ctx context.Context, ifaces config.Ifaces, enable, disable []string) {
 
-	cmCopy := cm.capturesCopy()
```

Comment:
I completely removed the need to pull a full copy of the captures all the time. Instead, the new type `Captures` cleanly wraps all accesses to the map with barriers / locks where needed. Make this code much leaner and safer (because there's no need to worry about this anymore).